cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,311 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+type RolloutStatus int
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+const (
+	// resource desired status is not upgraded yet
+	ToUpgrade RolloutStatus = iota
+	// resource desired status is upgraded and last applied status is not upgraded
+	Upgrading
+	// resource desired status is upgraded and last applied status is upgraded successfully
+	UpgradeSucceed
+	// resource desired status is upgraded and last applied status upgrade failed
+	UpgradeFailed
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// rollout status
+	status RolloutStatus
+	// the last transition time of rollout status
+	lastTransitionTime *time.Time
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// duck type rollout strategy
+	rolloutStrategy RolloutStrategy
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+	// Return the rollout status on each managed cluster
+	clusterRolloutStatusFunc ClusterRolloutStatusFunc
+}
+
+func NewRolloutHandler(rolloutStrategy RolloutStrategy, pdTracker *clusterv1beta1.PlacementDecisionClustersTracker, clusterRolloutStatusFunc ClusterRolloutStatusFunc) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		rolloutStrategy:          rolloutStrategy,
+		pdTracker:                pdTracker,
+		clusterRolloutStatusFunc: clusterRolloutStatusFunc,
+	}
+}
+
+// Return the strategy actual take effect and a list of cluster that needs to update with current state
```

Comment:
The RolloutHandler already have the RolloutStrategy struct that already has its rolloutType. Why do we have GetRolloutAllClusters, GetRolloutProgressive and GetRolloutProgressivePerGroup Functions as public ? 
GetRolloutCluster Func should be enough ?

Reply:
Yes, agree, modify them as private func.